libretro-rs/src/lib.rs

```diff
@@ -106,16 +107,27 @@ impl TryFrom<u32> for RetroDevice {
   }
 }
 
-#[derive(Clone, Copy, Debug)]
-pub struct RetroDevicePort(u32);
+/// A libretro device port.
+#[repr(transparent)]
+#[derive(Clone, Copy, Debug, PartialEq, Eq, PartialOrd, Ord)]
+pub struct RetroDevicePort(c_uint);
```

Comment:
Why the change to `c_uint`?

Reply:
`RetroInstance::on_set_controller_port_device` is going to create a RetroDevicePort out of a `c_uint` using the private constructor. Shouldn't we ensure ensure the types always match on every architecture?

Reply:
Translations of this kind should be done in the glue layer, everything user-facing should be a safe, known-size, rust type in my opinion. Using `c_uint` here would cause `RetroDevicePort` to be a different size on different architectures.

In a worse case, it might force us to expose our `libc::c_uint` if we aren't very careful about API design. I think a good "rule of thumb" here is that all translations from the C API to Rust happen in the `on_*` methods of `RetroInstance` (either directly, or by calling an appropriate conversion on a struct, say `RetroDevicePort::try_from_raw` before we pass that data along to the trait. The structs themselves, however, should just have `u32` or whatever equivalent rust type.

Reply:
Considering Rust doesn't have implicit numeric coercions like C or a stable ABI, is the architecture-specific size really a problem? Even safe Rust has to deal with architecture-specific `usize` right?

I think we're on the same page regarding hiding as much of the C API's ugliness as possible (e.g. no raw pointers, newtypes instead of plain integer types, enforce safety by design) but I don't see how we can work around the fact that various parts of the API are architecture-specific without surfacing that to the end user, especially in this case where the core's both sending and receiving port numbers.

The way I look at it, the `c_uint` is win/win. If `c_uint = u32` in every architecture, the change makes no difference. But if there's cases where `c_uint != u32`, sticking with `u32` will either force us to 1) cause a compilation error, 2) cast at runtime and pray we never cast a value that'd overflow, or 3) check for overflow and panic. All of these seem strictly worse than surfacing the architecture-specific size to the user. That said I've never had to write FFI code or maintain a crate before so it's possible there's something I've overlooked? I'm not trying to twist your arm, I genuinely want to understand what makes the fixed size approach better so I can be better informed.

As for `libc`, I think we can remove that dependency altogether. Bindgen isn't generating `libc` types in `libretro.rs`, it's using `std::os::raw`. Since the crate still compiles when mixing `libc` and `std::os::raw` types, they must be using the same type aliases? Actually, the crate and its users should probably both be using the `core::ffi` types, which were copied over from `std::os::raw` so they could be used in `no_std` environments. I believe the types in `std::os::raw` now alias the ones in `core::ffi`, and the documentation for `std::os::raw` tells you to prefer the `core::ffi` module instead.

Reply:
After sleeping on it (literally) I realized there's no problem using a `u8` after all. Since we're limiting cores to using ports 0-255, if the frontend calls `retro_set_controller_device_port` with port 256+, there's nothing useful the core can do with that info and we can just ignore the function call.